src/util/fs.js

```diff
@@ -251,6 +251,179 @@ async function buildActionsForCopy(
   }
 }
 
+async function buildActionsForHardlink(
+  queue: CopyQueue,
+  events: CopyOptions,
+  possibleExtraneous: Set<string>,
+  reporter: Reporter,
+): Promise<CopyActions> {
+  const artifactFiles: Set<string> = new Set(events.artifactFiles || []);
+  const files: Set<string> = new Set();
+
+  // initialise events
+  for (const item of queue) {
+    const onDone = item.onDone;
+    item.onDone = () => {
+      events.onProgress(item.dest);
+      if (onDone) {
+        onDone();
+      }
+    };
+  }
+  events.onStart(queue.length);
+
+  // start building actions
+  const actions: CopyActions = [];
+
+  // custom concurrency logic as we're always executing stacks of 4 queue items
+  // at a time due to the requirement to push items onto the queue
+  while (queue.length) {
+    const items = queue.splice(0, 4);
```

Comment:
Maybe pull out the 4 as a constant (`const CONCURRENT_QUEUE_ITEMS = 4` or something like that, that explains what it means)